modules/internal/ChapelArray.chpl

```diff
@@ -3988,49 +4016,119 @@ module ChapelArray {
   pragma "init copy fn"
   proc chpl__initCopy(ir: _iteratorRecord) {
 
-    // The use of an explicit initCopy() is required
-    // to support nested for/forall expressions.
-    iter _ir_copy_recursive(ir) {
-      for e in ir {
-        pragma "no copy"
-        var ee = chpl__initCopy(e);
+    // We'd like to know the yielded type of the record, but we can't
+    // access the (runtime) component of that until we actually yield
+    // something.
+    //
+    // The below is a work-around that creates a 1-D array by
+    // filling in a ddata and then handing the data to DefaultRectangularArr.
+    // Besides breaking the catch-22 with the runtime type, the other
+    // advantage of this approach is that it moves the yielded value
+    // in to the array instead of assigning it. Alternatives to this strategy
+    // include creating a no-inited array with the static type that the
+    // iterator yields.
+    //
+    // Of course, in the future, we'd like this to support
+    // creating multidimensional arrays - but that requires storing
+    // the yielded shape in the iterator record which doesn't seem close
+    // at hand...
+    //
+    // The resulting array grows dynamically. That wouldn't always be
+    // necessary if we had better shape information in iterator records.
+    // Additionally, the dynamic growing limits parallelism here, so
+    // adding better shape information would allow it to be parallel and
+    // thus enable better performance.
 
-        yield ee;
-      }
-    }
+    var i  = 0;
+    var size:size_t = 0;
+    var data:_ddata(iteratorToArrayElementType(ir.type)) = nil;
 
-    pragma "no copy"
-    var irc  = _ir_copy_recursive(ir);
+    var callAgain: bool;
+    var subloc = c_sublocid_none;
+
+    for elt in ir {
+
+      // Future: we should generally remove this copy.
+      // Note though that in some cases it invokes this function
+      // recursively - in that case it shouldn't be removed!
+      pragma "no auto destroy"
+      pragma "no copy"
+      var eltCopy = chpl__initCopy(elt);
+
+      pragma "insert auto destroy"
+      type eltType = eltCopy.type;
```

Comment:
I do not see you use `eltType`. If this is intentional, please add a brief comment explaining why. Otherwise please remove `eltType`.